typescript/infra/scripts/verify-validator.ts

```diff
@@ -104,170 +88,259 @@ class S3Wrapper {
-      `Missing checkpoints (${missingCheckpoints.length}): ${missingCheckpoints}\n`,
-    );
-  if (invalidCheckpoints.length)
-    console.log(
-      `Invalid checkpoints (${invalidCheckpoints.length}): ${invalidCheckpoints}\n`,
-    );
+  private async getLatestCheckpoints(): Promise<{
+    controlLatestCheckpoint: number;
+    prospectiveLastCheckpoint: number;
+  }> {
+    const [
+      { obj: controlLatestCheckpoint },
+      { obj: prospectiveLastCheckpoint },
+    ] = await Promise.all([
+      this.controlS3BucketClient
+        .getS3Obj<number>('checkpoint_latest_index.json')
+        .catch((err) => {
+          console.error(
+            "Failed to get control validator's latest checkpoint.",
+            err,
+          );
+          process.exit(1);
+        }),
+      this.prospectiveS3BucketClient
+        .getS3Obj<number>('checkpoint_latest_index.json')
+        .catch((err) => {
+          console.error(
+            "Failed to get prospective validator's latest checkpoint.",
+            err,
+          );
+          process.exit(1);
+        }),
+    ]);
 
-  if (modTimeDeltasS.length > 1) {
-    // Drop the time of the first one since it is probably way off
-    modTimeDeltasS.length--;
-    console.log(
-      `Time deltas (∆ < 0 -> prospective came earlier than the control)`,
-    );
-    console.log(modTimeDeltasS);
-    console.log(`Median: ${median(modTimeDeltasS)}s`);
-    console.log(`Mean:   ${mean(modTimeDeltasS)}s`);
-    console.log(`Stdev:  ${stdDev(modTimeDeltasS)}s`);
+    if (
+      !isLatestCheckpoint(controlLatestCheckpoint) ||
+      !isLatestCheckpoint(prospectiveLastCheckpoint)
+    )
+      throw new Error('Invalid latest checkpoint data');
+
+    console.log(`Latest Index`);
+    console.log(`control: ${controlLatestCheckpoint}`);
+    console.log(`prospective: ${prospectiveLastCheckpoint}\n`);
+    return { controlLatestCheckpoint, prospectiveLastCheckpoint };
+  }
+
+  private async getControlCheckpoint(
+    checkpointIndex: number,
+  ): Promise<{ control: Checkpoint; controlLastMod: Date } | null> {
+    let control: Checkpoint, controlLastMod: Date, unrecoverableError;
+    try {
+      const s3Object = await this.controlS3BucketClient.getS3Obj(
+        this.checkpointKey(checkpointIndex),
+      );
+      if (isCheckpoint(s3Object.obj)) {
+        if (s3Object.obj.checkpoint.index != checkpointIndex) {
+          console.log(`${checkpointIndex}: Control index is invalid`, s3Object);
+          process.exit(1);
+        }
+        [control, controlLastMod] = [s3Object.obj, s3Object.modified];
+      } else {
+        console.log(`${checkpointIndex}: Invalid control checkpoint`, s3Object);
+        unrecoverableError = new Error('Invalid control checkpoint.');
+      }
+    } catch (err) {
+      return null;
+    }
+    if (unrecoverableError) throw unrecoverableError;
+    return { control: control!, controlLastMod: controlLastMod! };
+  }
+
+  private async getProspectiveCheckpoint(
```

Comment:
this is pretty similar to `getControlCheckpoint`
can we dedupe?

Reply:
I wanted to, but because of the difference in logic around consequences of a control missing versus a prospective missing they are just different enough for it to be more messy than it is worth in my opinion. Better to just keep the duplicates together.